File: sop-sdk/sdk-c++/common/OpenClient.cpp
#include <json/json.h>
#include "OpenClient.h"
#include "httplib.h"
#include "tool.h"
#include "sign.h"

httplib::Headers headers = {
        {"Accept-Encoding", "identity"}
};

const string ERROR_NODE = "error_response";

OpenClient::OpenClient(const string &appId, const string &privateKeyFilePath, const string &url) {
    this->appId = appId;
    this->url = url;
    this->privateKeyFilePath = privateKeyFilePath;
}

Json::Value OpenClient::execute(BaseRequest *request) {
    return this->execute(request, "");
}

Json::Value OpenClient::execute(BaseRequest *request,  const string &token) {
    string method = request->getMethod();
    string version = request->getVersion();
    RequestType requestType = request->getRequestType();
    map<string, string> bizModel = request->bizModel;
    // 创建HTTP请求客户端
//    httplib::Client cli(this->hostInfo.host, this->hostInfo.port);
    httplib::Client cli(this->url.c_str());
    const char *url = this->url.c_str();
    // 构建请求参数
    map<string, string> allParams = this->buildParams(request, token);
    string responseBody;
    // 如果有文件上传
    if (!request->getFiles().empty()) {
        httplib::MultipartFormDataItems items = OpenClient::getMultipartFormDataItems(
                allParams, request->getFiles());
        responseBody = cli.Post(url, headers, items)->body;
    } else {
        switch (requestType) {
            case GET: {
                string fullPath = this->url + "?" + OpenClient::getQuery(allParams);
                responseBody = cli.Get(fullPath.c_str())->body;
                break;
            }
            case POST_FORM: {
                responseBody = cli.Post(url, headers, OpenClient::getParams(allParams))->body;
                break;
            }
            case POST_JSON: {
                string json = tool::mapToJson(allParams);
                responseBody = cli.Post(url, json, "application/json")->body;
                break;
            }
            case POST_FILE: {
                httplib::MultipartFormDataItems items = OpenClient::getMultipartFormDataItems(
                        allParams, request->getFiles());
                responseBody = cli.Post(url, headers, items)->body;
            }
        }
    }
    return OpenClient::parseResponse(responseBody, request);
}


httplib::Params OpenClient::getParams(map<string, string> allParams) {
    httplib::Params params;
    map<string, string>::iterator it;
    for (it = allParams.begin(); it != allParams.end(); ++it) {
        params.emplace(it->first, it->second);
    }
    return params;
}

string OpenClient::getQuery(map<string, string> allParams) {
    string query;
    map<string, string>::iterator it;
    int i = 0;
    for (it = allParams.begin(); it != allParams.end(); ++it) {
        if (i++ > 0) {
            query.append("&");
        }
        query.append(it->first).append("=").append(tool::url_encode(it->second));
    }
    return query;
}

map<string, string> OpenClient::buildParams(BaseRequest *request, const string &token) {
    map<string, string> allParams;
    allParams["app_id"] = this->appId;
    allParams["method"] = request->getMethod();
    allParams["charset"] = "UTF-8";
    allParams["sign_type"] = "RSA2";
    allParams["timestamp"] = tool::getTime();
    allParams["version"] = request->getVersion();
    allParams["biz_content"] = tool::mapToJson(request->bizModel);

    if (!token.empty()) {
        allParams["app_auth_token"] = token;
    }

    // 生成签名
    string sign = signutil::createSign(allParams, this->privateKeyFilePath, "RSA2");
    allParams["sign"] = sign;
    return allParams;
}

httplib::MultipartFormDataItems
OpenClient::getMultipartFormDataItems(map<string, string> allParams, vector<FileInfo> fileInfoList) {
    httplib::MultipartFormDataItems items = {};
    map<string, string>::iterator it;
    for (it = allParams.begin(); it != allParams.end(); ++it) {
        items.push_back({it->first, it->second, "", ""});
    }
    // 添加上传文件
    vector<FileInfo>::iterator vit;
    for (vit = fileInfoList.begin(); vit != fileInfoList.end(); vit++) {
        string content = tool::getFileContent(vit->filepath);
        items.push_back({vit->name, content, tool::getFilename(vit->filepath), "application/octet-stream"});
    }
    return items;
}

Json::Value OpenClient::parseResponse(const string& responseBody, BaseRequest *request) {
    Json::Value root;
    Json::Reader reader;
    reader.parse(responseBody, root, false);

    Json::Value data = root[ERROR_NODE];
    if (!data) {
        string method = request->getMethod();
        string nodeName = tool::replace(method.c_str(),".","_") + "_response";
        data = root[nodeName];
    }
    return data;
}

File: sop-sdk/sdk-nodejs/common/OpenClient.js
const axios = require('axios');
const moment = require('moment');
const qs = require('qs');

const RequestType = require('./RequestType');
const SignUtil = require('./SignUtil');
const BaseRequest = require('./BaseRequest');

const IS_RUN_IN_BROWSER = typeof window !== 'undefined' && this === window;

const HEADERS = {'Accept-Encoding': 'identity'};

const getHeaders = (headers = {}) => {
    return Object.assign({}, headers, HEADERS);
};

const parseResponse = (error, response, request) => {
    if (!error && response.status === 200) {
        return request.parseResponse(response.data);
    } else {
        return { // 重新封装请求异常回调，以防中断
            msg: '请求异常',
            code: '502',
            sub_msg: `${error}`,
            sub_code: 'isv.invalid-server'
        };
    }
};

const buildParams = (instance, request, token) => {
    const {appId, privateKey} = instance;
    const allParams = {
        'app_id': appId,
        'method': request.getMethod(),
        'charset': 'UTF-8',
        'sign_type': 'RSA2',
        'timestamp': moment().format('YYYY-MM-DD HH:mm:ss'),
        'version': request.getVersion(),
        'biz_content': JSON.stringify(request.bizModel)
    };

    if (token) {
        allParams['app_auth_token'] = token;
    }
    // 创建签名
    allParams.sign = SignUtil.createSign(allParams, privateKey, 'RSA2');
    return allParams;
};

const executeRequest = async (instance = {}, request, token, callback, customOptions = {}) => {
    const params = buildParams(instance, request, token);
    const {url} = instance;
    let {headers} = customOptions;
    const config = {
        url,
        method: 'POST',
        params: undefined,
        data: undefined
    };
    headers = getHeaders(headers);
    const requestType = request.getRealRequestType();
    switch (requestType) {
        case RequestType.GET: {
            config.method = 'GET';
            config.params = params;
        }
            break;
        case RequestType.POST_FORM: {
            headers = Object.assign(headers, {
                'Content-Type': 'application/x-www-form-urlencoded'
            });
            config.data = qs.stringify(params);
        }
            break;
        case RequestType.POST_JSON: {
            config.data = params;
        }
            break;
        case RequestType.POST_FILE: {
            let formData;
            if (IS_RUN_IN_BROWSER) {
                formData = new window.FormData();
                (request.files || []).forEach(({name, path}) => {
                    formData.append(name, path, {
                        contentType: 'application/octet-stream'
                    });
                });
            } else {
                const fs = require('fs');
                const fd = require('form-data');
                formData = new fd();
                (request.files || []).forEach(({name, path}) => {
                    formData.append(name, fs.createReadStream(path), {
                        contentType: 'application/octet-stream'
                    });
                });
                headers = Object.assign(headers, formData.getHeaders());
            }
            Object.keys(params).forEach(key => {
                const value = params[key];
                if (!(typeof key === 'undefined' || typeof value === 'undefined')) {
                    formData.append(key, params[key]);
                }
            });
            config.data = formData;
        }
            break;
        default: {
            callback(parseResponse(new Error('request.getRequestType()类型不正确'), undefined, request));
            return;
        }
    }
    try {
        config['headers'] = headers;
        const response = await axios.request(config);
        callback(parseResponse(undefined, response, request));
    } catch (error) {
        callback(parseResponse(error, undefined, request));
    }
};

module.exports = class OpenClient {
    /**
     * 初始化客户端
     * @param appId 应用ID
     * @param privateKey 应用私钥，2048位，PKCS8
     * @param url 请求url
     */
    constructor(appId, privateKey, url) {
        this.appId = appId || '';
        this.privateKey = privateKey || '';
        this.url = url || '';
    }

    setAppId(appId) {
        this.appId = appId || '';
        return this;
    }

    setPrivateKey(privateKey) {
        this.privateKey = privateKey || '';
        return this;
    }

    setUrl(url) {
        this.url = url || '';
        return this;
    }

    /**
     * 发送请求
     * @param request 请求类
     * @param callback 回调函数，参数json（undefined则使用executeSync）
     * @param options 自定义参数，如headers
     */
    execute(request, callback, options) {
        if (typeof callback == 'function') {
            return this.executeToken(request, null, callback, options);
        } else {
            return this.executeSync(request, options);
        }
    }

    /**
     * 发送同步请求
     * @param request 请求类
     * @param options 自定义参数，如headers
     * */
    executeSync(request, options) {
        return new Promise(async (resolve) => {
            await this.execute(request, res => {
                resolve(res);
            }, options);
        });
    }

    /**
     * 发送请求
     * @param request 请求类
     * @param token token
     * @param callback 回调函数，参数json（undefined则使用executeTokenSync）
     * @param options 自定义参数，如headers
     */
    async executeToken(request, token, callback, options) {
        if (!(request instanceof BaseRequest)) {
            throw 'request类未继承BaseRequest';
        }
        if (typeof callback == 'function') {
            const files = request.files;
            if (files && files.length > 0) {
                request.setForceRequestType(RequestType.POST_FILE);
            }
            return await executeRequest(this, request, token, callback, options);
        } else {
            return this.executeTokenSync(request, token);
        }
    }

    /**
     * 发送同步请求
     * @param request 请求类
     * @param token token
     * @param options 自定义参数，如headers
     */
    executeTokenSync(request, token, options) {
        return new Promise(async (resolve) => {
            await this.executeToken(request, token, res => {
                resolve(res);
            }, options);
        });
    }

};
